proposals/interface-types/Explainer.md

```diff
@@ -252,36 +252,59 @@ resulting string? For example:
-`std::string` or `std::unique_ptr<char[]>`.
+This situation would naturally arise when exporting, e.g., a C++ function that
+returns a `std::string` or `std::shared_ptr<char[]>`. In both cases, the
+returned C++ object essentially contains a pointer along with the assumption
+that the calling C++ code is responsible for calling the object's destructor,
+which then frees the memory or drops a reference count, resp. But if that
+function is exported from an adapted wasm module, the caller isn't C++ code,
+it's the adapter function.
 
-What we need is to be able to call `free` right after the bytes are
-read from `memory-to-string`. Thus, `memory-to-string` takes an
-optional exported function name that it calls after it has read the string:
+What we need is to be able to call `free` (or some other export corresponding
+to a destructor) after the linear memory bytes are read by `memory-to-string`.
+Because there is no limitation on the number of `call-export`s in an adapter
+function, a first attempt at a solution would be to just call `free` directly:
 
 ```wasm
   (@interface func (export "greeting") (result string)
     call-export "greeting_"
-    memory-to-string "mem" "free"
+    dup
+    memory-to-string "mem"
+    swap
+    call-export "free"
   )
 ```
 
-Note that the ability to call *any* function also allows reference counting
-schemes to call a decrement function, e.g., if a C++ function returned a
-`std::shared_ptr<std::string>`.
+Here we use classic stack manipulation operations to duplicate the `i32` pointer
+value so that it can be read by both `memory-to-string` and the call to `free`.
+Unfortunately, there is a problem: anticipating the [exception handling]
+proposal, if an exception is thrown between the `call-export "greeting_"` and
+the `call-export "free"`, the memory will be leaked. With more complicated
+signatures, there can be many instructions in this range which can throw.
+
+To address this problem, as well as another problem described [below](#shared-nothing-linking-example),
+there is a `defer-call-export` instruction:
+
+```wasm
+  (@interface func (export "greeting") (result string)
+    call-export "greeting_"
+    defer-call-export "free"
+    memory-to-string "mem"
+  )
+```
 
-> **TODO** Proper integration with exception handling will likely require
-> switching to a different scheme that is exception safe, so that an exception
-> thrown between `call "greeting_"` and `memory-to-string` *also* calls
-> `"free"`. Moreover, the call to `"free"` should happen after not just
-> `memory-to-string`, but after the `string` value is consumed by an
-> adapter instruction, so that the pair of instructions can be optimized
-> into a direct linear memory copy.
+This instruction requests that the given function (`free`) be called on all
+exits from the current frame (normal and exceptional). The arguments for
+this deferred call are taken from the top of the stack at the time of the
```

Comment:
nit: maybe "copied from the top of the stack" to make it extra-clear that it's more tee-like?

Reply:
sgtm, fixed!